src/OrchardCore.Modules/OrchardCore.Menu/MenuShapes.cs

```diff
@@ -102,8 +135,39 @@ public override ValueTask DiscoverAsync(ShapeTableBuilder builder)
 
                 if (menuItems != null)
                 {
+                    IEnumerable<Permission> allPermissions = null;
+
+                    var permissionService = context.ServiceProvider.GetRequiredService<IPermissionService>();
+                    var httpContextAccessor = context.ServiceProvider.GetRequiredService<IHttpContextAccessor>();
+                    var authorizationService = context.ServiceProvider.GetRequiredService<IAuthorizationService>();
+
                     foreach (var contentItem in menuItems)
                     {
+                        var hasPermission = true;
+
+                        if (contentItem.TryGet<MenuItemPermissionPart>(out var permissionPart) &&
+                        permissionPart.PermissionNames is not null &&
+                        permissionPart.PermissionNames.Length > 0)
+                        {
```

Comment:
lazily initialize the services

`permissionService ??= context.ServiceProvider.GetRequiredService<IPermissionService>();`

Reply:
check out the changes.